appstore/validator.go

```diff
@@ -114,13 +110,40 @@ func (c *Client) Verify(req IAPRequest, result interface{}) error {
 	b := new(bytes.Buffer)
 	json.NewEncoder(b).Encode(req)
 
-	resp, err := client.Post(c.URL, "application/json; charset=utf-8", b)
+	resp, err := client.Post(c.ProductionURL, "application/json; charset=utf-8", b)
 	if err != nil {
 		return err
 	}
 	defer resp.Body.Close()
 
-	err = json.NewDecoder(resp.Body).Decode(result)
+	// Read the body now so that we can unmarshal it twice
+	buf, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	err = json.Unmarshal(buf, &result)
+	if err != nil {
+		return err
+	}
+
+	// https://developer.apple.com/library/content/technotes/tn2413/_index.html#//apple_ref/doc/uid/DTS40016228-CH1-RECEIPTURL
+	var r StatusResponse
+	err = json.Unmarshal(buf, &r)
+	if err != nil {
+		return err
+	}
+	if r.Status == 21007 {
+		b = new(bytes.Buffer)
+		json.NewEncoder(b).Encode(req)
+		resp, err := client.Post(c.SandboxURL, "application/json; charset=utf-8", b)
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+
+		return json.NewDecoder(resp.Body).Decode(result)
+	}
 
-	return err
+	return nil
```

Comment:
Actually I am a bit confused about the usecase of this Verify() method now.
Can ```err``` be ```nil``` when the verification is actually not successful?

Reply:
err in this case means that we encountered an error during the process of verifying the receipt, whether it be failed connection or some bad response. It does not however mean that the receipt is not valid. I guess there could be cases where non 200 status codes are returned and gets passed up (don't quite remember and I'm away from my laptop). Though you can definitely feel free open an issue for this.

Reply:
Right, this is going to be a separate issue. This does not support the case well, when you only want to verify the receipt and do not care about the details.